Fix missing lint output for organizeDeclarations rule

calda opened the pull request:
This PR fixes an issue where lint violations for the `organizeDeclarations` rule would never be printed when using `--lint`. This could create a confusing experience where the command would output "Source input did not pass lint check" but not tell you while rule the violations came from.

calda commented on Tests/CommandLineTests.swift:
```diff
@@ -658,4 +658,46 @@ class CommandLineTests: XCTestCase {
+
+        _ = processArguments(["", "stdin", "--lint", "--rules", "organizeDeclarations"], in: "")
+
+        XCTAssertEqual(output, [
+            "Running SwiftFormat...",
+            "(lint mode - no files will be changed.)",
+            ":2:1: error: (organizeDeclarations) Organize declarations within class, struct, enum, actor, and extension bodies.",
+            ":3:1: error: (organizeDeclarations) Organize declarations within class, struct, enum, actor, and extension bodies.",
+            ":4:1: error: (organizeDeclarations) Organize declarations within class, struct, enum, actor, and extension bodies.",
+            ":8:1: error: (organizeDeclarations) Organize declarations within class, struct, enum, actor, and extension bodies.",
+            "Source input did not pass lint check.",
+        ])
```
Before the changes in this PR this test only output:

```
Running SwiftFormat...
(lint mode - no files will be changed.)
Source input did not pass lint check.
```

calda commented on Sources/SwiftFormat.swift:
```diff
@@ -567,18 +567,40 @@ private func applyRules(
+            // If all of the changes got filtered out, it's because the only changes are _moves_
+            // that don't modify any individual lines. We still want to emit some list of changes
+            // in this case, so soften the filtering to only exclude duplicates.
+            if !changes.isEmpty, filteredChanges.isEmpty {
```
The `organizeDeclarations` rule sometimes just moves code without actually modifying the content of the lines. The diffing implementation above was filtering those out. Instead we can use more relaxed filtering implementation when necessary.

calda commented:
Actually this causes other test failures that I'm not sure how to fix, so I'm just going to close this for now.

nicklockwood commented:
Perhaps what's needed is a more sophisticated model for change tracking, where we distinguish between edits and moves? Then if the change type was "edit" but the line hasn't changed then it wouldn't be logged, but if it's "move" then it still would be?

calda commented:
Thanks for the suggestion @nicklockwood, that sort of approach worked much better!

I added a new `formatter.moveTokens` helper that moves tokens and tags the `Formatter.Change` as a move that shouldn't be filtered out later. Since `organizeDeclarations` works by replacing `formatter.tokens` all at once, I also added a `formatter.updateAllTokens` helper that diffs the lines and tags the `Formatter.Change` for any moved lines as a move.

I also added a check for this issue in the `testFormatting` helper, which highlighted that the `docCommentsBeforeAttributes` rule also had this issue. Changing that rule to use `formatter.moveTokens` fixed it.

nicklockwood commented:
Amazing! Thanks - lgtm
